design/pool-configuration.md

```diff
@@ -0,0 +1,552 @@
+metadata:
+  name: bgpadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    peers:
+      - peer1
+      - peer2
+    - communities:
+        - 65535:65282
+    aggregationLength: 32
+    localPref: 100
+```
+
+A `BGPAdvertisement` with no pool selector is intended to select all the available `AddressPool`s.
+
+#### L2 Advertisement
+
+The `L2Advertisement` entity represents the intention to advertise a given IP via L2. Currently, there are no knobs related to L2 advertisement.
+
+```go
+type l2Advertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: lsadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+```
+
+A `L2Advertisement` with no pool selector is intended to select all the available `AddressPool`s.
+
+### BGP Peer
+
+The `BgpPeer` entity represents an external BGP peer we want to establish a session with. It contains the properties of the sessions, and it's a 1:1 mapping of the current bgp peer structure without the node selector. We also add the `bfdProfile` structure here, as it is part of the session definition.
+
+```go
+type bgpPeer struct {
+  Name              string
+  MyASN         uint32
+  ASN           uint32
+  Addr          string
+  SrcAddr       string
+  Port          uint16
+  HoldTime      string
+  KeepaliveTime string
+  RouterID      string
+  Password      string
+  BFDProfile    string
+  EBGPMultiHop  bool
+}
+
+type bfdProfile struct {
+  Name             string  
+  ReceiveInterval  *uint32 
+  TransmitInterval *uint32 
+  DetectMultiplier *uint32 
+  EchoInterval     *uint32 
+  EchoMode         bool    
+  PassiveMode      bool    
+  MinimumTTL       *uint32 
+}
+```
+
+Example CRDs:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPPeer
+metadata:
+  name: peer-bfd
+  namespace: metallb-system
+spec:
+  peerAddress: 10.0.0.2
+  peerASN: 64501
+  myASN: 64500
+  routerID: 10.10.10.10
+  peerPort: 1
+  holdTime: "10s"
+  sourceAddress: 1.1.1.1
+  bfdProfile: bfdprofilefull
+
+apiVersion: metallb.io/v1beta1
+kind: BFDProfile
+metadata:
+  name: bfdprofilefull
+  namespace: metallb-system
+spec:
+  receiveInterval: 35
+  transmitInterval: 35
+  detectMultiplier: 37
+  echoReceiveInterval: 38
+  echoTransmitInterval: 39
+  echoMode: true
+  passiveMode: true
+  minimumTtl: 10
+```  
+
+### New configuration degrees
+
+#### Announcing the IP from a subset of nodes
+
+The two Advertisement entities are extended with a selector of the nodes to announce from:
+
+A configuration would look like:
+
+```go
+type bgpAdvertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
+  AggregationLength *int
+  LocalPref         *uint32
+  Communities       []string
+  NodeSelector      *metav1.LabelSelector
```

Comment:
My bad, the peer selector was already in my mind (see the CR example above):
```
```yaml
apiVersion: metallb.io/v1beta1
kind: BGPAdvertisement
metadata:
  name: bgpadvertisement
  namespace: metallb-system
spec:
    pools:
      - pool1
      - pool2
    peers:
      - peer1
      - peer2
 ```

With your version, we won't be able to express the case where we want to advertise only some ips from a subset of the nodes.
That's because the BGP Peer must be unique (per peer) while we can have multiple bgpAdvertisements (and I notice now the peer part is missing from this struct, need to amend that).

Reply:
Also, the config is simpler if the focus is advertising to different peers from different nodes.
The most common use case I can envision is to have a pool advertised to a subset of peers from a subset of nodes, which will be simpler with the current proposed version, compared to having to add a nodeselector to each peer.

Reply:
As a second thought, I think we need the node selector BOTH on the bgp advertisement and on the bgp peer, because they express two different concepts.

The BGPPeer object expresses the desire to establish a BGP connection, the BGPAdvertisement expresses the desire to advertise a prefix via BGP. We need both because we may either want to connect to peers from a subset of nodes, and orthogonally we may want to advertise prefixes only from a subset of nodes.

The first one is on par with what we have right now, and even if we could get away with not having the selector in the bgppeer, that would result in wasted resources of metallb attempting to connect to a peer from a node that won't be able to reach it.